File: rust/problem_20.rs
// Copyright 2014 Mitchell Kember. Subject to the MIT License.
// Project Euler: Problem 20
// Factorial digit sum

use problem_13::big;
use problem_13::num::bigint::BigInt;
use problem_16::sum_of_digits;

fn factorial(n: int) -> BigInt {
	let mut fac = big(n);
	for i in range(1, n) {
		fac = fac * big(i);
	}
	fac
}

pub fn solve() -> int {
	sum_of_digits(factorial(100))
}

File: c++/src/problem_20.cpp
// Copyright 2016 Mitchell Kember. Subject to the MIT License.
// Project Euler: Problem 20
// Factorial digit sum

#include "common.hpp"

#include <gmpxx.h>

namespace problem_20 {

long solve() {
	mpz_class n(1);
	for (long i = 2; i <= 100; ++i) {
		n *= i;
	}
	return common::sum_of_digits(n);
}

} // namespace problem_20
